mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -83,8 +121,38 @@ fn match_struct(ast: &syn::DeriveInput, variant: &syn::VariantData) -> quote::To
             }
         }
 
+        #(#getters)*
+
+        #(#setters)*
+
+        #(#js_getters)*
+
+        #(#js_setters)*
+
+        lazy_static! {
+            pub static ref PS_ARR: [JSPropertySpec; #spec_size] = [
+                #(#js_prop_specs,)*
+                JSPropertySpec::end_spec(),
+            ];
+        }
+
         #(#slot_num)*
 
+        pub fn check_this(cx: *mut JSContext, args: &JS::CallArgs) -> Option<*mut JSObject> {
+            let jsobj = unsafe {
+                rooted!(in(cx) let thisv = args.thisv().get());
+                if !thisv.is_object() {
+                    return None;
+                }
+                let jsobj = thisv.to_object();
+                if jsapi::JS_GetClass(jsobj) as *const _ as usize != &#js_class as *const _ as usize {
+                    return None;
+                }
+                jsobj
+            };
+            Some(jsobj)
```

Comment:
This should probably be an unsafe function since it is using a raw pointer. I'd also tighten it up a little bit, like this:

```rust
pub unsafe fn check_this(cx: *mut jsapi::JSContext, args: &jsapi::JS::CallArgs) -> Option<#name> {
    let thisv = args.thisv().get();
    if !thisv.is_object() {
        return None;
    }
    #name::from_object(thisv.to_object())
}
```

I think this function could also make sense as a static method of `#name`, so it would be invoked as `#name::check_this(...)`